File: python/Brunel/_connection.py

__all__ = ["Connection"]


def _setState(state, val, default=None):
    try:
        result = state[val]
        if result:
            return result
        else:
            return default
    except KeyError:
        return default


def _mergeWeights(state, other, key):
    state = state[key]
    other = other[key]

    for key, value in other.items():
        if key not in state:
            state[key] = value
        else:
            state[key] += value


def _mergeSources(state, other, key):
    state = state[key]
    other = other[key]

    for key, values in other.items():
        if key not in state:
            state[key] = values
        else:
            for value in values:
                if value not in state[key]:
                    state[key].append(value)


class Connection:
    """Holds information about a Connection in the network"""
    def __init__(self, props=None, getHook=None):
        self._getHook = getHook

        self.state = {
            "id": None,
            "n0": None,
            "n1": None,
            "shared": None,
            "type": None,
            "duration": None,
            "affiliations": None,
            "correspondences": None,
            "projects": None,
            "weights": None,
            "notes": None,
        }

        self.setState(props)

    def setState(self, state):
        if not state:
            return

        self.state["id"] = _setState(state, "id")
        self.state["n0"] = _setState(state, "n0")
        self.state["n1"] = _setState(state, "n1")
        self.state["shared"] = _setState(state, "shared", [])
        self.state["type"] = _setState(state, "type")
        self.state["duration"] = _setState(state, "duration")
        self.state["affiliations"] = _setState(state, "affiliations", {})
        self.state["correspondences"] = _setState(state, "correspondences", {})
        self.state["notes"] = _setState(state, "notes", [])
        self.state["projects"] = _setState(state, "projects", {})
        self.state["weights"] = _setState(state, "weights", {})

        if self.state["n0"] is None or self.state["n1"] is None:
            raise ValueError(f"Invalid connection! {self.state['n0']}<=>"
                             f"{self.state['n1']} : {self.state}")

        if self.state["n0"] > self.state["n1"]:
            tmp = self.state["n0"]
            self.state["n0"] = self.state["n1"]
            self.state["n1"] = tmp

    def merge(self, other):
        if not self._getHook:
            if other._getHook:
                return other.merge(self)

        import copy as _copy
        state = _copy.deepcopy(self.state)

        for s in other.state["shared"]:
            if s not in state["shared"]:
                state["shared"].append(s)

        for n in other.state["notes"]:
            state["notes"].append(n)

        _mergeSources(state, other.state, "affiliations")
        _mergeSources(state, other.state, "correspondences")
        _mergeWeights(state, other.state, "weights")

        for id, project in other.state["projects"].items():
            state["projects"][id] = project

        state["duration"] = state["duration"].merge(other.state["duration"])

        m = Connection()
        m.state = state
        m._getHook = self._getHook

        return m

    def __str__(self):
        n0 = self.getNode0()
        n1 = self.getNode1()

        if n0 is None or n1 is None:
            return "Connection::null"
        else:
            return f"Connections({self.getNode0Name()} " \
                   f"<=> {self.getNode1Name()})"

    def __repr__(self):
        return self.__str__()

    @staticmethod
    def getConnectionName(node1, node2):
        n0 = node1.getID()
        n1 = node2.getID()

        if n0 <= n1:
            return f"{n0}<=>{n1}"
        else:
            return f"{n1}<=>{n0}"

    def getName(self):
        n0 = self.state["n0"]
        n1 = self.state["n1"]

        if n0 <= n1:
            return f"{n0}<=>{n1}"
        else:
            return f"{n1}<=>{n0}"

    def getNode0Name(self):
        node0 = self.getNode0()

        try:
            return node0.getName()
        except Exception:
            return node0

    def getNode1Name(self):
        node1 = self.getNode1()

        try:
            return node1.getName()
        except Exception:
            return node1

    def getNode0(self):
        node0 = self.state["n0"]

        if self._getHook:
            node0 = self._getHook(node0)

        return node0

    def getNode1(self):
        node1 = self.state["n1"]

        if self._getHook:
            node1 = self._getHook(node1)

        return node1

    def getDuration(self):
        return self.state["duration"]

    def getAffiliationSources(self):
        return self.state["affiliations"]

    def getcorrespondencesources(self):
        return self.state["correspondences"]

    def getID(self):
        return self.state["id"]

    def getWeights(self):
        return self.state["weights"]

    def getProjects(self):
        """ Get the projects this connection is a member of

            Note: The daterange values returned are null

            Returns:
                dict: Dictionary of project_id: Daterange pairs
        """
        return self.state["projects"]

    def inProject(self, project_id):
        """ Is this connection a part of the project with project_id

            Returns:
                bool: True if part of project else False
        """
        return project_id in self.state["projects"]

    def toDry(self):
        state = {}

        for key, value in self.state.items():
            if value:
                try:
                    if len(value) > 0:
                        state[key] = value
                except Exception:
                    state[key] = value

        return state

    @staticmethod
    def unDry(value):
        return Connection(value)

    @staticmethod
    def load(data):
        return Connection(data)

File: src/model/Connection.js
import Dry from "json-dry";
import lodash from "lodash";

import DateRange from "./DateRange";

import get_id from "./get_id";

import { ValueError } from "./Errors";

function setState(val, def = null) {
  if (val) {
    return val;
  } else {
    return def;
  }
}

function _mergeSources(state, other, key) {
  state = state[key];
  other = other[key];

  Object.keys(other).forEach((key) => {
    if (!(key in state)) {
      state[key] = other[key];
    } else {
      for (let item in other[key]) {
        if (!(item in state[key])) {
          state[key].push(item);
        }
      }
    }
  });
}

class Connection {
  constructor(props) {
    this.state = {
      id: null,
      n0: null,
      n1: null,
      shared: null,
      type: null,
      duration: null,
      affiliations: null,
      correspondances: null,
      projects: null,
      notes: null,
      is_highlighted: false,
    };

    this.setState(props);
    this._getHook = null;
    this._isAConnectionObject = true;
  }

  static clone(item) {
    let c = new Connection();
    c._getHook = item._getHook;
    c.state = lodash.cloneDeep(item.state);
    return c;
  }

  _updateHooks(hook) {
    this._getHook = hook;
  }

  getID() {
    return this.state.id;
  }

  getName() {
    let n0 = this.state.n0;
    let n1 = this.state.n1;

    if (n0 <= n1) {
      return `${n0}<=>${n1}`;
    } else {
      return `${n1}<=>${n0}`;
    }
  }

  getNode0() {
    if (this._getHook) {
      return this._getHook(this.state.n0);
    } else {
      return this.state.n0;
    }
  }

  getNode1() {
    if (this._getHook) {
      return this._getHook(this.state.n1);
    } else {
      return this.state.n1;
    }
  }

  getNode0ID() {
    return this.state.n0;
  }

  getNode1ID() {
    return this.state.n1;
  }

  areNodesVisible(n) {
    return n[this.state.n0] && n[this.state.n1];
  }

  getNode0Name() {
    let node0 = this.getNode0();

    try {
      return node0.getName();
    } catch (error) {
      return node0;
    }
  }

  getNode1Name() {
    let node1 = this.getNode1();

    try {
      return node1.getName();
    } catch (error) {
      return node1;
    }
  }

  getDuration() {
    return this.state.duration;
  }

  getAffiliationSources() {
    //console.log(this.state.affiliations);
    return this.state.affiliations;
  }

  getCorrespondanceSources() {
    //console.log(this.state.correspondances);
    return this.state.correspondances;
  }

  filterSource(source) {
    if (source.getID) {
      let id = source.getID();
      source = {};
      source[id] = 1;
    }

    let nsources = Object.keys(source).length;

    let seen = {};

    Object.keys(source).forEach((key) => {
      if (this.state.affiliations[key] || this.state.correspondances[key]) {
        seen[key] = 1;
      }
    });

    if (Object.keys(seen).length !== nsources) {
      return null;
    } else {
      return this;
    }
  }

  filterProject(project) {
    // Filter this connection by project ID
    if (project.getID) {
      let id = project.getID();
      project = {};
      project[id] = 1;
    }

    let nProjects = Object.keys(project).length;

    let seen = {};

    Object.keys(this.state.projects).forEach((key) => {
      if (key in project) {
        seen[key] = 1;
      }
    });

    if (Object.keys(seen).length !== nProjects) {
      return null;
    } else {
      return this;
    }
  }

  filterWindow(window) {
    if (!window) {
      return this;
    } else if (!window._isADateRangeObject) {
      window = new DateRange(window);
    }

    if (!window) {
      return this;
    }

    let duration = this.getDuration();

    if (!duration) {
      return this;
    }

    let intersect = window.intersect(duration);

    if (!intersect) {
      return null;
    } else {
      return this;
    }
  }

  setState(state) {
    if (state) {
      this.state.id = setState(state.id);
      this.state.n0 = setState(state.n0);
      this.state.n1 = setState(state.n1);
      this.state.shared = setState(state.shared, []);
      this.state.type = setState(state.type);
      this.state.duration = setState(state.duration);
      this.state.affiliations = setState(state.affiliations, {});
      // note that correspondences is a typo from the python!
      this.state.correspondances = setState(state.correspondences, {});
      this.state.notes = setState(state.notes, []);
      this.state.weights = setState(state.weights, {});
      this.state.projects = setState(state.projects, {});

      if (!this.state.n0 || !this.state.n1) {
        throw new ValueError(`Invalid connection ${this}`);
      }
    }
  }

  merge(other) {
    if (!this._getHook) {
      if (other._getHook) {
        return other.merge(this);
      }
    }

    let state = lodash.cloneDeep(this.state);

    Object.keys(other.state.shared).forEach((key) => {
      if (!(key in state.shared)) {
        state.shared[key] = other.state.shared[key];
      }
    });

    Object.keys(other.state.notes).forEach((key) => {
      if (!(key in state.notes)) {
        state.notes[key] = other.state.notes[key];
      }
    });

    _mergeSources(state, other.state, "affiliations");
    _mergeSources(state, other.state, "correspondances");

    state.duration = state.duration.merge(other.state.duration);

    let c = new Connection();
    c.state = state;
    c._getHook = this._getHook;

    return c;
  }

  toString() {
    return `Connection(${this.getNode0Name()}<=>${this.getNode1Name()})`;
  }

  getColorFromType() {
    switch (this.state.type) {
      case "direct":
        return "rgb(180,150,150)";
      case "indirect":
        return "rgb(180,180,180)";
      default:
        return "orange";
    }
  }

  getType() {
    return this.state.type;
  }

  getWeights() {
    return this.state.weights;
  }

  getWeight() {
    let weight = 1.0;

    Object.keys(this.state.weights).forEach((key) => {
      weight += this.state.weights[key];
    });

    return weight;
  }

  getWeightFromType() {
    let weight = this.getWeight();

    switch (this.state.type) {
      case "direct":
        return 1.0 * weight;
      case "indirect":
        return 0.5 * weight;
      default:
        console.log(`Unknown type? ${this.state.type}`);
        return 0.5 * weight;
    }
  }

  setHighlighted(val) {
    if (val) {
      this.state.is_highlighted = true;
    }
    else {
      this.state.is_highlighted = false;
    }
  }

  getHighlighted() {
    return this.state.is_highlighted;
  }

  toEdge() {
    let color = this.getColorFromType();
    // let weight = this.getWeightFromType();
    let weight = this.getWeight();

    let edge = {
      id: this.getID(),
      source: get_id(this.state.n0),
      target: get_id(this.state.n1),
      value: weight,
      type: this.state.type,
      color: color,
      highlighted: this.getHighlighted(),
    };

    return edge;
  }

  toDry() {
    return { value: this.state };
  }
}

Connection.unDry = function (value) {
  return new Connection(value);
};

Connection.load = function (data) {
  return new Connection({ name: data.name });
};

Dry.registerClass("Connection", Connection);

export default Connection;
